To emulate the outcomes of the Gearbox system, we instantiate a network comprising 5000 nodes, with a portion represented by $P_a$ being designated as adversarial. These adversarial nodes are distributed randomly within the network. Nodes are subsequently assigned to shards through a random allocation process. We then assess each shard to determine whether the percentage of adversarial nodes exceeds the shard's predefined liveness threshold. In cases where this threshold is surpassed, we dissolve the existing shard and construct a new one with a higher gear setting. This process is iteratively executed until all shards maintain a percentage of adversarial nodes that falls below their respective liveness thresholds. There are five gears corresponding to the liveness threshold of $10\%$, $20\%$, $25\%$, $30\%$ and $49\%$. We repeat this entire procedure 1000 times, observing the distribution of shards across different gear configurations.


### 1. Simulating the average size and overlap time of all shards in Gearbox under different adversary ratios

Total node count: 5000
Gearbox: L=10%, 20%, 25%, 30%, and 49%
Corresponding shard sizes: 26, 39, 50, 63, 293

In [2]:
import random

"""The simulation about Gearbox:
    We generated 5000 nodes and there are five gears, gears size are 26, 39, 50, 63, 293. We calculate it from the Shard size.ipynb with failure rate falling in 10^-7 and 10^-8.
    We assign nodes into shards randomly and check if the adversarial nodes in the shard exceed this shard's liveliness threshold or if we need to move to a larger one.
    We keep doing this until all the shards can run normally and then calculate the average shard size. Additionally, we calculate the overlap time.
"""

"""Function run_simulation used to find the average shard size of Gearbox with different adversarial nodes ratios.
    Input:
        n: The upper bound of the adversarial nodes ratio in the system.
        
    Output:
        Average shard size of all shards.
        
"""

# Function run_simulation used for finding the average shard size of Gearbox with different adversarial nodes ratios.
def run_simulation(pa):
    # Calculate the number of Byzantine nodes based on the percentage 'Pa'
    byzantine_nodes = int(pa / 100 * 5000)
    results = []
    element_values = []  # Used to store the element values for each run
    for i in range(192):
        # Step 1: Select a sample of 26 nodes. There are 192 shards each size 26 nodes.
        sample_1 = random.sample(range(5000), 26)
        num_byzantine_nodes = sum(1 for node in sample_1 if node < byzantine_nodes)  #We assume the Node index 0~byzantine_node belongs to adversary. 
        if num_byzantine_nodes > 2:
            sample_2 = random.sample(range(5000), 39)
            num_byzantine_nodes = sum(1 for node in sample_2 if node < byzantine_nodes)
            if num_byzantine_nodes > 7:
                sample_3 = random.sample(range(5000), 50)
                num_byzantine_nodes = sum(1 for node in sample_3 if node < byzantine_nodes)
                if num_byzantine_nodes > 12:
                    sample_4 = random.sample(range(5000), 63)
                    num_byzantine_nodes = sum(1 for node in sample_4 if node < byzantine_nodes)
                    if num_byzantine_nodes > 18:
                        element = 293
                    else:
                        element = 63
                else:
                    element = 50
            else:
                element = 39
        else:
            element = 26
        
        results.append(element)
        element_values.append(element)  # Collect element values for each run

    # Calculate the average number of records
    return sum(results) / len(results), element_values

gresult = []
goverlap = []

for pa in range(50):
    results = []
    element_results = []

    for i in range(1000):
        # Run the simulation for 'pa' and obtain the average number of records and element values
        avg_records, element_values = run_simulation(pa)
        results.append(avg_records)
        element_results.extend(element_values)  # Add element values for each run to element_results

    # Calculate the percentage of each element
    total_elements = len(element_results)
    element_percentages = {26: 0, 39: 0, 50: 0, 63: 0, 293: 0}

    for element in element_results:
        element_percentages[element] += 1

    # Divide the percentages by the total to ensure they sum up to 100%
    for key in element_percentages:
        element_percentages[key] /= total_elements

    # Output the percentages for each element
    print(f"pa={pa}:")
    for key, percentage in element_percentages.items():
        print(f"{key}: {percentage * 100:.2f}%")

    gresult.append(sum(results) / len(results))
    print(f"Pa={pa}: {gresult[-1]}")
    # Calculate the overlap time by dividing the average results by 26
    print(f"overlap time: {gresult[-1] / 26}")
    goverlap.append(gresult[-1] / 26)


pa=0:
26: 100.00%
39: 0.00%
50: 0.00%
63: 0.00%
293: 0.00%
Pa=0: 26.0
overlap time: 1.0
pa=1:
26: 99.77%
39: 0.23%
50: 0.00%
63: 0.00%
293: 0.00%
Pa=1: 26.029723958333264
overlap time: 1.001143229166664
pa=2:
26: 98.53%
39: 1.47%
50: 0.00%
63: 0.00%
293: 0.00%
Pa=2: 26.191614583333312
overlap time: 1.0073697916666657
pa=3:
26: 95.85%
39: 4.15%
50: 0.00%
63: 0.00%
293: 0.00%
Pa=3: 26.5393645833333
overlap time: 1.0207447916666654
pa=4:
26: 91.70%
39: 8.30%
50: 0.00%
63: 0.00%
293: 0.00%
Pa=4: 27.079510416666714
overlap time: 1.0415196314102582
pa=5:
26: 86.27%
39: 13.72%
50: 0.01%
63: 0.00%
293: 0.00%
Pa=5: 27.785312499999986
overlap time: 1.0686658653846148
pa=6:
26: 79.78%
39: 20.18%
50: 0.03%
63: 0.00%
293: 0.00%
Pa=6: 28.631901041666673
overlap time: 1.1012269631410259
pa=7:
26: 72.69%
39: 27.18%
50: 0.13%
63: 0.00%
293: 0.00%
Pa=7: 29.56441666666671
overlap time: 1.1370929487179504
pa=8:
26: 65.30%
39: 34.33%
50: 0.37%
63: 0.00%
293: 0.00%
Pa=8: 30.55163020833334
overlap time: 1.17

In our paper, we also take into account the latency incurred when disseminating data, and we model this latency as linearly proportional to the number of nodes involved. To ensure the time-bound for shard-related activities is greater than the latency incurred during message transmission, we've established a relationship that satisfies $l = 2.12x$ where $x$ represents the number of nodes within the shard. Note that this estimation does not consider the potential slowdown when a node is in multiple shards and have multiple workload in parallel.  $2.12$ is chosen base on the fact that Rapidchain would be able to generate a block and conduct one round of voting that involves $329$ nodes in a shard within $698s$, which is measured in runtime in the same experiment environment for Reticulum and Rapidchain. Consequently, the time-bound for block generation and the subsequent voting rounds are set at $55.12s$, $82.68$, $106s$, $133.56s$, and $621.16s$, corresponding to the various shard sizes in different gears. This approach ensures that the time intervals are both consistent with the Gearbox paper's principles and suitably adjusted to account for the latency incurred in our specific context.


## 2. Calculate the throughput of the Gearbox at different adversary ratios

2.1 When $P_a=10%$:

26: 50.78%

39: 47.50%

50: 1.72%

63: 0.00%

293: 0.00%

overlap time: 1.2533935897435906

50.78% of the shards reached consensus within 26 nodes, and they took a total time of: 26 * 2.12

47.50% of the shards reached consensus within 39 nodes, and they took a total time of: 39 * 2.12

1.72% of the shards reached consensus within 50 nodes, and they took a total time of: 50 * 2.12

Total consensus time: 106 seconds

Throughput = (5000 / 26) * 4096 / 106 / 1.253 = 5930.61 transactions per second

2.2 When $P_a=20%:
26: 8.32%
39: 42.41%
50: 40.27%
63: 8.63%
293: 0.37%

overlap time: 1.7445641025641037

8.32% of the shards reached consensus within 26 nodes, and they took a total time of: 26 * 2.12

42.41% of the shards reached consensus within 39 nodes, and they took a total time of: 39 * 2.12

40.27% of the shards reached consensus within 50 nodes, and they took a total time of: 50 * 2.12

8.63% of the shards reached consensus within 63 nodes, and they took a total time of: 63 * 2.12

Total average consensus time: 133.56 seconds

Throughput = (5000 / 26) * 4096 / 133.56 / 1.744 = 3381.68 transactions per second

3.3 When $P_a=30%:

26: 0.66%
39: 6.53%
50: 20.38%
63: 33.35%
293: 39.08%
overlap time: 5.7089576923077

0.66% of the shards reached consensus within 26 nodes, and they took a total time of: 26 * 2.12

6.53% of the shards reached consensus within 39 nodes, and they took a total time of: 39 * 2.12

20.38% of the shards reached consensus within 50 nodes, and they took a total time of: 50 * 2.12

33.35% of the shards reached consensus within 63 nodes, and they took a total time of: 63 * 2.12

39.08% of the shards reached consensus within 293 nodes, and they took a total time of: 293 * 2.12

Total consensus time: 621.16 seconds

Throughput = (5000 / 26) * 4096 / 621.16 / 5.708 = 222.16 transactions per second